webapi/Skills/ChatSkills/ChatSkill.cs

```diff
@@ -394,59 +446,42 @@ private async Task<ChatMessage> GetChatResponseAsync(string chatId, string userI
+            chatContextTokenCount += TokenUtilities.GetContextMessageTokenCount(AuthorRole.System, contextComponent);
+        }
 
+        // Fill in the chat history with remaining token budget.
         string chatHistory = string.Empty;
+        var chatHistoryTokenBudget = remainingTokenBudget - chatContextTokenCount - TokenUtilities.GetContextMessageTokenCount(AuthorRole.System, planResult);
 
-        // Append the chat history, if allowed.
-        if (chatHistoryTokenLimit > 0)
-        {
-            await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting chat history", cancellationToken);
-            chatHistory = await this.ExtractChatHistoryAsync(chatId, chatHistoryTokenLimit, cancellationToken);
-            chatContext.ThrowIfFailed();
-            chatContextText = $"{chatContextText}\n{chatHistory}";
-        }
+        // Append previous messages
+        await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting chat history", cancellationToken);
+        chatHistory = await this.GetAllowedChatHistoryAsync(chatId, chatHistoryTokenBudget, promptTemplate, cancellationToken);
+        chatContext.ThrowIfFailed();
 
         // Append the plan result last, if exists, to imply precedence.
         if (!string.IsNullOrWhiteSpace(planResult))
         {
-            chatContextText = $"{chatContextText}\n{planResult}";
+            promptTemplate.AddSystemMessage(planResult);
         }
 
-        // Set variables needed in prompt
-        chatContext.Variables.Set("audience", audience);
-        chatContext.Variables.Set("userIntent", userIntent);
-        chatContext.Variables.Set("chatContext", chatContextText);
-
-        // Render the prompt
-        var promptRenderer = new PromptTemplateEngine();
-        var renderedPrompt = await promptRenderer.RenderAsync(
-            this._promptOptions.SystemChatPrompt,
-            chatContext,
-            cancellationToken);
-        chatContext.Variables.Set("prompt", renderedPrompt);
-
-        // Need to extract this from the rendered prompt because Time and Date are calculated during render
-        var systemChatContinuation = Regex.Match(renderedPrompt, PromptsOptions.SYSTEM_CHAT_CONTINUATION_REGEX).Value;
-        var promptView = new BotResponsePrompt(renderedPrompt, this._promptOptions.SystemDescription, this._promptOptions.SystemResponse, audience, userIntent, chatMemories, documentMemories, plannerDetails, chatHistory, systemChatContinuation);
-
-        // Calculate token usage of prompt template
-        chatContext.Variables.Set(TokenUtilities.GetFunctionKey(this._logger, "SystemMetaPrompt")!, TokenUtilities.TokenCount(renderedPrompt).ToString(CultureInfo.InvariantCulture));
-
+        var promptView = new BotResponsePrompt(this._promptOptions.SystemDescription, this._promptOptions.SystemResponse, audience, userIntent, chatMemories, documentMemories, plannerDetails, chatHistory, promptTemplate);
```

Comment:
Looks like the prompt template is the object that's passed to the model and it contains everything the BotResponsePrompt contains. Why do we still need the BosResponsePrompt object?

Reply:
To show the formatted sections in the UX. Otherwise, there's not a clean way to determine each component of the prompt. The prompt template is shown as the raw content
![image](https://github.com/microsoft/chat-copilot/assets/125500434/51683ce7-184f-4ba5-b94a-eac3bd8ab15b)

Reply:
Got it!